### PR TITLE
Serialize ValueCache

### LINKED ISSUE #26695: Persistent ValueCache
## Reason
The `ValueCache` object is a kd-tree based cache object that can store arbitrary data associated with points in an n-dimensional space (e.g. speciation and chemical potentials from a Gibbs energy minimization as a function of state space coordinates such as concentrations and temperature). We should add functionality have this cache persist between runs to make the most out of the previously computed data previously computed.

## Design
Provide a file name constructor argument. The file is read upon construction and updated upon destruction of the cache object.

## Impact
Persistent caching of data that is expensive to compute.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #26695

This adds `DataIO` support to `ValueCache`. That means value caches can now be restartable, or with the new constructor, persistent between runs.

### CODE REVIEW

File: framework/include/utils/ValueCache.h [hunk 1]
```diff
@@ -62,34 +79,72 @@ class ValueCache
...
+    dataLoad(in_file, *this, nullptr);
+
+    // build kd-tree
+    _kd_tree = std::make_unique<KdTreeT>(_in_dim, _cache, _max_leaf_size);
+    mooseAssert(_kd_tree != nullptr, "KDTree was not properly initialized.");
+  }
+}
+
+template <typename T>
+ValueCache<T>::~ValueCache()
```
Comment (loganharbour) [inline framework/include/utils/ValueCache.h:128]: When do these get destructed in actual use? Storing to file in a destructor sounds like it _could_ be dangerous

Comment (dschwen) [reply framework/include/utils/ValueCache.h:128]: Nah, destructors are guaranteed to fully run upon application exit. If you need the cache to survive unplanned crashed or HPC wall time expiry you need to make it restartable, which should now work, too.

File: framework/doc/content/source/utils/ValueCache.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+When the object is constructed using the
+
+```C++
+ValueCache(const std::string & file_name, std::size_t in_dim, std::size_t max_leaf_size = 10);
+```
+
+constructor, the state of the cache is saved to a file upon destuction of the cache object (usually at teh ond of a simulation), and read in at construction of the cache object. Thus teh cached knowledge from a previous simulation will be available in future runs.
+
+!alert warning
+It is up to the developer to ensure that the cache state from previous runs is applicable to the following runs when teh persistence feature is used!
```
Comment (loganharbour) [inline framework/doc/content/source/utils/ValueCache.md:16]:
```suggestion
It is up to the developer to ensure that the cache state from previous runs is applicable to the following runs when the persistence feature is used!
```

File: framework/include/utils/ValueCache.h [hunk 3]
```diff
@@ -28,32 +41,36 @@ class ValueCache
 {
 public:
   ValueCache(std::size_t in_dim, std::size_t max_leaf_size = 10);
+  ValueCache(const std::string & file_name, std::size_t in_dim, std::size_t max_leaf_size = 10);
```
Comment (loganharbour) [inline framework/include/utils/ValueCache.h:44]: Can you please add just a little docstring action here

**Review by loganharbour — CHANGES_REQUESTED** [review]
Good otherwise once green.

You are the "teh" master

File: framework/include/utils/ValueCache.h (hunk 3 above)
Comment (dschwen) [reply framework/include/utils/ValueCache.h:48]: done

**Comment by dschwen** [comment]
@loganharbour I did some more refactoring after your review.

**Review by loganharbour — APPROVED** [review]

**Comment by dschwen** [comment]
@loganharbour after a quick discussion we decided to change the API a bit.

**Review by loganharbour — APPROVED** [review]

**Pull request merged**